record-layer.go

```diff
@@ -42,72 +51,111 @@ type RecordLayer struct {
 	cachedRecord *TLSPlaintext // Last record read, cached to enable "peek"
 	cachedError  error         // Error on the last record read
 
-	ivLength int         // Length of the seq and nonce fields
-	seq      []byte      // Zero-padded sequence number
-	nonce    []byte      // Buffer for per-record nonces
-	cipher   cipher.AEAD // AEAD cipher
+	cipher   *cipherState
+	datagram bool
 }
 
-type recordLayerFrameDetails struct{}
+type recordLayerFrameDetails struct {
+	datagram bool
+}
 
 func (d recordLayerFrameDetails) headerLen() int {
-	return recordHeaderLen
+	if d.datagram {
+		return recordHeaderLenDtls
+	}
+	return recordHeaderLenTls
 }
 
 func (d recordLayerFrameDetails) defaultReadLen() int {
-	return recordHeaderLen + maxFragmentLen
+	return d.headerLen() + maxFragmentLen
 }
 
 func (d recordLayerFrameDetails) frameLen(hdr []byte) (int, error) {
-	return (int(hdr[3]) << 8) | int(hdr[4]), nil
+	return (int(hdr[d.headerLen()-2]) << 8) | int(hdr[d.headerLen()-1]), nil
 }
 
-func NewRecordLayer(conn io.ReadWriter) *RecordLayer {
+func newCipherStateNull() *cipherState {
```

Comment:
Is this needed?  Why not just use `nil` as the null `*cipherState`?  We already switch encryption/decryption on whether `cipher != nil`.

Reply:
You need it to store the sequence number